*: update dependencies (for gccgo)

tamird opened the pull request:
Enables building on gccgo 8.0 (pre-release), though linker flags will
require additional massaging, see
https://groups.google.com/forum/#!msg/golang-nuts/SlGCPYkjxo4/4DjcjXRCqAkJ

Release note: None

Vendored PR: https://github.com/cockroachdb/vendored/pull/14

EDIT: with these changes `GODEBUG=cgocheck=0 make test` (https://github.com/golang/go/issues/23391) passes all tests on gccgo except `storage`'s `verifyBelowRaftProtos` which fails with
```
*enginepb.MVCCMetadata: not observed below raft!
*enginepb.MVCCStats: not observed below raft!
*raftpb.HardState: not observed below raft!
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21477)
<!-- Reviewable:end -->

dt commented:
I'm a little wary of the across-the-board dep bump. At least I would feel a little better if it were easier to see what changed... maybe include a `dep-diff-parser` summary in the commit/PR to make it easier to audit those?

tamird commented:
Ah, forgot about that tool. Added to commit message, though I'm not sure it's going to make it much easier to reason about which ones are important and which ones aren't.

tamird commented:
Some notes:
- random-looking import/grpc API changes are from when I did the dependency bump before @bdarnell's https://github.com/cockroachdb/cockroach/pull/21398 (my version allocates less :man_shrugging:)
- a few dependencies are here moved to "stable" releases (rather than their master branches), which should reduce future bump frequency
- generated protobuf changed inconsequentially
- generated stringer code changed inconsequentially
- raft introduced a new error upstream where no error was previously signaled - it is here coerced to `nil` with TODOs added

tamird committed:
Revert "sql: work around lack of runtime/trace in gccgo"

This reverts commit ce6c652e8f4355d5a8ee90aebdd7e129f1b53300.

gccgo now supports runtime/trace.

tamird committed:
storage/engine: gccgo compatibility

runtime.rawbyteslice is inlined in gccgo, so mallocgc is used instead.

name                                                     old time/op    new time/op    delta
MVCCScan_RocksDB/rows=100/versions=1/valueSize=64-16       33.4µs ± 0%    33.0µs ± 2%  -1.10%  (p=0.003 n=8+10)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=512-16      62.1µs ± 3%    61.6µs ± 1%    ~     (p=0.165 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-16       270µs ± 2%     271µs ± 2%    ~     (p=0.842 n=9+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-16      491µs ± 2%     506µs ± 3%  +3.04%  (p=0.001 n=10+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=64-16     3.02ms ± 1%    3.02ms ± 3%    ~     (p=0.515 n=8+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=512-16    10.8ms ± 4%    10.7ms ± 4%    ~     (p=0.393 n=10+10)

name                                                     old speed      new speed      delta
MVCCScan_RocksDB/rows=100/versions=1/valueSize=64-16      192MB/s ± 0%   194MB/s ± 2%  +1.12%  (p=0.003 n=8+10)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=512-16     824MB/s ± 3%   832MB/s ± 1%    ~     (p=0.165 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-16     237MB/s ± 2%   236MB/s ± 2%    ~     (p=0.860 n=9+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-16   1.04GB/s ± 2%  1.01GB/s ± 3%  -2.94%  (p=0.001 n=10+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=64-16    212MB/s ± 1%   212MB/s ± 2%    ~     (p=0.515 n=8+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=512-16   473MB/s ± 4%   478MB/s ± 4%    ~     (p=0.393 n=10+10)

tamird committed:
build: permit gccgo version which includes gcc version

e.g. `go version go1.10rc1 gccgo (GCC) 8.0.1 20180131 (experimental) linux/amd64`

Release note: None

tamird committed:
*: update minimal dependencies for gccgo support

Also:
 - Update golang/dep
 - Enable pruning (removes ~6M LOC from vendor)
 - Move all possibile dependencies to tagged releases

This commit enables building on gccgo 8.0 (pre-release), though linker
flags will require additional massaging, see
https://groups.google.com/forum/#!msg/golang-nuts/SlGCPYkjxo4/4DjcjXRCqAkJ

$ dep-diff-parser
Changed:
 - github.com/apache/thrift: a533ea16 -> 327ebb6c
   - https://github.com/apache/thrift/compare/a533ea16...327ebb6c
 - github.com/axiomhq/hyperloglog: 83009472 -> baba800b
   - https://github.com/axiomhq/hyperloglog/compare/83009472...baba800b
 - github.com/dgryski/go-metro: 0f647357 -> 280f6062
   - https://github.com/dgryski/go-metro/compare/0f647357...280f6062
 - github.com/go-ole/go-ole: 2ecd927e -> a41e3c4b
   - https://github.com/go-ole/go-ole/compare/2ecd927e...a41e3c4b
 - github.com/gogo/protobuf: fd9a4790 -> 1adfc126
   - https://github.com/gogo/protobuf/compare/fd9a4790...1adfc126
 - github.com/golang/dep: 8ddfc8af -> 37d9ea0a
   - https://github.com/golang/dep/compare/8ddfc8af...37d9ea0a
 - github.com/golang/protobuf: 1643683e -> bbd03ef6
   - https://github.com/golang/protobuf/compare/1643683e...bbd03ef6
 - github.com/google/go-github: 996760c5 -> e48060a2
   - https://github.com/google/go-github/compare/996760c5...e48060a2
 - github.com/grpc-ecosystem/grpc-gateway: 8879672d -> 07f5e797
   - https://github.com/grpc-ecosystem/grpc-gateway/compare/8879672d...07f5e797
 - github.com/jmank88/nuts: a1e02c78 -> 8b28145d
   - https://github.com/jmank88/nuts/compare/a1e02c78...8b28145d
 - github.com/mattn/goveralls: a32275e3 -> b71a1e48
   - https://github.com/mattn/goveralls/compare/a32275e3...b71a1e48
 - github.com/openzipkin/zipkin-go-opentracing: 9c88fa03 -> 45e90b00
   - https://github.com/openzipkin/zipkin-go-opentracing/compare/9c88fa03...45e90b00
 - github.com/pelletier/go-toml: 4e9e0ee1 -> acdc4509
   - https://github.com/pelletier/go-toml/compare/4e9e0ee1...acdc4509
 - github.com/petermattis/goid: 3db12ebb -> b0b1615b
   - https://github.com/petermattis/goid/compare/3db12ebb...b0b1615b
 - github.com/sasha-s/go-deadlock: 565eb443 -> 03d40e5d
   - https://github.com/sasha-s/go-deadlock/compare/565eb443...03d40e5d
 - github.com/satori/go.uuid: 5bf94b69 -> f58768cc
   - https://github.com/satori/go.uuid/compare/5bf94b69...f58768cc
 - honnef.co/go/tools: 80dfe556 -> d73ab98e
   - https://github.com/dominikh/go-tools/compare/80dfe556...d73ab98e
Removed:
 - github.com/dgryski/go-bits

Release note: None

tamird committed:
cli: avoid runtime.SetCgoSymbolizer on gccgo

It is not supported.

Release note: None

tamird committed:
rpc: remove errant comment

This was missed in 6f97c89.

Release note: None

tamird committed:
*: rearrange imports

Release note: None

tamird committed:
cli,rpc: avoid some allocations

Release note: None

tamird committed:
*: avoid order of evaluation issues in gccgo

See https://github.com/golang/go/issues/23188.

Release note: None

tamird committed:
server/debug: simplify

See https://golang.org/cl/71251.

Release note: None

tamird committed:
sql/pgwire: add go1.10 test shim

Release note: None

tamird committed:
*: improve style, fix inaccurate error message

Release note: None

tamird committed:
*: work around gccgo inconsistencies

Release note: None

tamird committed:
ts: improve readability of test failures

Release note: None

tamird committed:
ts: avoid NPE in test

Release note: None

tamird committed:
sql/logictest: consistent spacing

Release note: None

tamird committed:
*: make tests compatible with go1.10

Release note: None

benesch reviewed (commented):
This all LGTM, but I think @bdarnell needs to sign off on any dep bump this late in the cycle. (Using pinned releases where possible seems like an important goal, though.)

tamird commented on pkg/storage/engine/slice.go:
```diff
@@ -0,0 +1,39 @@
+	src := (*[maxArrayLen]byte)(ptr)[:len:len]
+	copy(x, src)
+	return x
+}
```
Go 1.11 is a long way away; not doing this would prevent experimenting with gccgo until August.

benesch replied:
Oh, sorry, I didn't read this thoroughly enough. I thought this file was support for Go < 1.9. This seems fine!

bdarnell reviewed (approved):
Thanks for taking the time to trim this down to the minimal set of updates!

tamird commented:
Thanks for reviewing!

tamird commented:
Can you merge the vendored PR?

petermattis commented:
Review status: 0 of 73 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/engine/slice.go, line 25 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/21477#-L4cZkmWp2TirbteUFO6:-L4cZkmWp2TirbteUFO7:baxcoko) ([raw file](https://github.com/cockroachdb/cockroach/blob/8594cb3de483e2b5c892c45db81737c55af98946/pkg/storage/engine/slice.go#L25)):*
> ```Go
> 	var b []byte
> 	*(*reflect.SliceHeader)(unsafe.Pointer(&b)) = reflect.SliceHeader{
> 		Data: uintptr(mallocgc(uintptr(len), nil, false)),
> ```

This isn't safe. A `uintptr` does not act as a reference to memory as far as the GC is concerned. `reflect.SliceHeader` is usually a bug because of that. Between the call to `mallocgc` and the eventual assignment to `b` the memory could be collected. Instead, you should be using the same `maxArrayLen` trick that is used below in `gobytes`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21477)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 73 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/engine/slice.go, line 25 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/21477#-L4cZkmWp2TirbteUFO6:-L4chNa5HAcsMVeirSfc:b4lrehl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8594cb3de483e2b5c892c45db81737c55af98946/pkg/storage/engine/slice.go#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This isn't safe. A `uintptr` does not act as a reference to memory as far as the GC is concerned. `reflect.SliceHeader` is usually a bug because of that. Between the call to `mallocgc` and the eventual assignment to `b` the memory could be collected. Instead, you should be using the same `maxArrayLen` trick that is used below in `gobytes`. 
</blockquote></details>

Thanks for taking a look! I did not understand why the `maxArrayLen` dance was being done below. https://github.com/cockroachdb/cockroach/pull/22400 implements your suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21477)*
<!-- Sent from Reviewable.io -->
